clippy_lints/src/redundant_locals.rs

```diff
@@ -0,0 +1,100 @@
+use clippy_utils::{diagnostics::span_lint_and_help, is_from_proc_macro, ty::needs_ordered_drop};
+use rustc_hir::{def::Res, BindingAnnotation, ByRef, Expr, ExprKind, HirId, Local, Node, Pat, PatKind, QPath};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::Ident;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for redundant redefinitions of local bindings.
+    ///
+    /// ### Why is this bad?
+    /// Redundant redefinitions of local bindings do not change behavior and are likely to be unintended.
+    ///
+    /// Note that although these bindings do not affect your code's meaning, they _may_ affect `rustc`'s stack allocation.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let a = 0;
+    /// let a = a;
+    ///
+    /// fn foo(b: i32) {
+    ///    let b = b;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let a = 0;
+    /// // no redefinition with the same name
+    ///
+    /// fn foo(b: i32) {
+    ///   // no redefinition with the same name
+    /// }
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub REDUNDANT_LOCALS,
+    correctness,
```

Comment:
Is this really a `correctness` lint? What's the danger in doing this? deny-by-default for this lint seems a bit strict to me.

I would rather put this lint into `complexity`. Maybe `suspicious`.

Reply:
I put it in the correctness category because `self_assignment`, which is almost the same thing, is also in correctness.

Reply:
`self_assignment` is a lot worse (it causes logic bugs) but yeah, I think this should be `correctness` because it's never what the programmer intended

Reply:
Thanks for running `lintcheck`. In that case, I think it can keep `correctness`. I guess "code that is outright useless" fits for this lint.